papa/src/main/java/papa/internal/RealInputTracker.kt

```diff
@@ -23,49 +24,84 @@ import papa.InputTracker
 import papa.SafeTrace
 import papa.internal.FrozenFrameOnTouchDetector.findPressedView
 import papa.safeTrace
+import java.util.concurrent.TimeUnit
+import kotlin.time.Duration.Companion.nanoseconds
 
 internal object RealInputTracker : InputTracker {
 
   override val motionEventTriggeringClick: DeliveredInput<MotionEvent>?
-    get() = motionEventTriggeringClickLocal.get()
+    get() = motionEventTriggeringClickLocal.get()?.input
 
   override val currentKeyEvent: DeliveredInput<KeyEvent>?
     get() = currentKeyEventLocal.get()
 
-  private val motionEventTriggeringClickLocal = ThreadLocal<DeliveredInput<MotionEvent>>()
+  private val motionEventTriggeringClickLocal = ThreadLocal<MotionEventHolder>()
```

Comment:
One thing to note: the thread locals here aren't in charge of actually holding the event ( / event holder) for the duration of its lifecycle, but instead of holding the event in specific time spans (wrapping onClick callback invocations) so that we can reach back here from an onClick and capture the event but also ensure that if we reach back outside of an onClick sandwich we actually find nothing event if overall the EventHolder is still hanging out in memory.

Reply:
(and of course this should actually be a comment in the code)